client/lxd.go

```diff
@@ -492,3 +493,196 @@ func (r *ProtocolLXD) getUnderlyingHTTPTransport() (*http.Transport, error) {
+			return req, nil
+		case api.ContainerSnapshotPost:
+			req.Target.Operation = fmt.Sprintf("%s/1.0/operations/%s", serverURL, neturl.PathEscape(opURL))
+			return req, nil
+		case api.StorageVolumePost:
+			req.Target.Operation = fmt.Sprintf("%s/1.0/operations/%s", serverURL, neturl.PathEscape(opURL))
+			return req, nil
+		default:
+			return nil, fmt.Errorf("Unknown request type")
+		}
+	}
+
+	getErrMsgForReq := func(req any) string {
+		switch req.(type) {
+		case api.InstancesPost:
+			return "Failed instance creation"
+		case api.InstanceRebuildPost:
+			return "Failed instance rebuild"
+		case api.ContainersPost:
+			return "Failed container creation"
+		case api.ImagesPost:
+			return "Failed remote image download"
+		case api.StorageVolumesPost:
+			return "Failed storage volume creation"
+		case api.InstancePost:
+			return "Failed instance migration"
+		case api.InstanceSnapshotPost:
+			return "Failed instance migration"
+		case api.ContainerPost:
+			return "Failed container migration"
+		case api.ContainerSnapshotPost:
+			return "Failed container migration"
+		case api.StorageVolumePost:
+			return "Failed storage volume creation"
+		default:
+			return "Unknown request type"
+		}
+	}
+
+	reflectedType := reflect.ValueOf(typeIface)
+	reflectedF := reflectedType.MethodByName(methodName)
+	success := false
+	var errors []remoteOperationResult
+	var op Operation
+	for _, serverURL := range urls {
+		sourcedReq, err := setSourceOrTarget(serverURL)
+		if err != nil {
+			errors = append(errors, remoteOperationResult{URL: serverURL, Error: err})
+			continue
+		}
+
+		params := make([]reflect.Value, 0)
+		if len(reqAttr) > 0 {
+			for _, attr := range reqAttr {
+				params = append(params, reflect.ValueOf(attr))
+			}
+		}
+
+		params = append(params, reflect.ValueOf(sourcedReq))
+		if len(optArgs) > 0 {
+			// args other than reqAttr come last as their are optional.
+			for _, arg := range optArgs {
+				params = append(params, reflect.ValueOf(arg))
+			}
+		}
+
+		out := reflectedF.Call(params)
+		op = out[0].Interface().(Operation)
+		err = out[1].Interface().(error)
+		if err != nil {
+			errors = append(errors, remoteOperationResult{URL: serverURL, Error: err})
+			continue
+		}
+
+		rop.handlerLock.Lock()
+		rop.targetOp = op
+		rop.handlerLock.Unlock()
+
+		for _, handler := range rop.handlers {
+			_, _ = rop.targetOp.AddHandler(handler)
+		}
+
+		err = rop.targetOp.Wait()
+		if err != nil {
+			errors = append(errors, remoteOperationResult{URL: serverURL, Error: err})
+			if shared.IsConnectionError(err) {
+				continue
+			}
+
+			break
+		}
+
+		success = true
+		break
+	}
+
+	if !success {
+		rop.err = remoteOperationError(getErrMsgForReq(req), errors)
+		if op != nil {
+			_ = op.Cancel()
+		}
+	}
+
+	close(rop.chDone)
+}
+
+func (r *ProtocolLXD) optimizeSrcImage(source ImageServer, image api.Image, instSrc *api.InstanceSource) (isSameServer bool, info *ConnectionInfo, err error) {
```

Comment:
I think this function makes sense, but lets have this as a separate series of commits, before anything to do with instance rebuild is added.